# Numbers and more in Python!

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 

Throughout this code we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

 
 
Now let's start with some basic arithmetic.

### Basic Arithmetic

In [2]:
print('checking type of whole num',type(3*2))
print(type(4))

checking type of whole num <class 'int'>
<class 'int'>


In [3]:
type(4.55)

float

In [9]:
# integer variable
abc = 100
print(type(abc))
print(abc)

<class 'int'>
100


In [4]:
# float variable
f_1 = 12.14
type(f_1)

float

In [12]:
a = 3+2j
type(a)

complex

In [13]:
abc = complex(2,4)
print(type(abc))
print(abc)

<class 'complex'>
(2+4j)


In [20]:
# Subtraction
2-1


1

In [6]:
# Multiplication
2*2

4

In [14]:
# floor
33//12

# module
33%2

2

In [8]:
4/2

2.0

In [9]:
7/4

1.75

In [10]:
# Floor Division
7//4

1

**Whoa! What just happened? .... 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

In [3]:
from math import floor
a = 10
b = 3

print(floor(a/b))  # 3

3


In [12]:
floor(10/3)

3

**So what if we just want the remainder after division?**

In [13]:
# Modulo
7%4

3

4 goes into 7 once, with a remainder of 3. The % operator returns the remainder after division.

### Arithmetic continued

In [22]:
# Powers
2**3
pow(2,3)

8

In [25]:
import math 
# Can also do roots this way
4**0.5
math.sqrt(4)

2.0

In [16]:
# Order of Operations followed in Python 
2 + 10 * 10 + 3

105

(PEMDAS=Parentheses, Exponents, Multiplication/Division, Addition/Subtraction)

In [1]:
# Can use parentheses to specify orders
(2+10) * (10+3)* 2**2

624

## Variable Assignments

Now that we've seen how to use numbers in Python as a calculator let's see how we can assign names and create variables.

We use a single equals sign to assign labels to variables. Let's see a few examples of how we can do this.

In [3]:
# Let's create an object called "a" and assign it the number 5
# variable can be used to store any datatype
# Datatypes 1. Integer 2. Float 3. Complex
a = 5
b = 7.5
c = 4+7j
print(type(a))
print(type(b))
print(type(c))

<class 'int'>
<class 'float'>
<class 'complex'>


In [19]:
print("type of a = ",type(a))
print("type of b= ",type(b))
print("type of addition",type(a+b))

type of a =  <class 'int'>
type of b=  <class 'float'>
type of addition <class 'float'>


Now if I call *a* in my Python script, Python will treat it as the number 5.

In [20]:
# Adding the objects
a+b

12.5

What happens on reassignment? Will Python let us write it over?

In [21]:
# Reassignment
a = 10.6

In [22]:
# Check
a

10.6

Yes! Python allows you to write over assigned variable names. We can also use the variables themselves when doing the reassignment. Here is an example of what I mean:

In [23]:
# Check
a

10.6

In [24]:
# Use A to redefine A
a = a + a

In [25]:
# Check 
a

21.2

The names you use when creating these labels need to follow a few rules:

    1. Names can not start with a number.
    2. There can be no spaces in the name, use underscore ('_') instead.
    3. Can't use any of these symbols :'",<>/?|\()!@#$%^&*~-+
    4. It's considered best practice (PEP8) that names are lowercase.
    5. Avoid using the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), or 'I' (uppercase letter eye) as single character variable names.
    6. Avoid using words that have special meaning in Python like "list" and "str"


Using variable names can be a very useful way to keep track of different variables in Python. For example:

In [26]:
print1 = 4 
print

<function print>

In [27]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate

In [29]:
Name = input("Enter Name")
print(Name)

Enter Name gadfadhf


gadfadhf


In [5]:
c = complex(2,3)
c

(2+3j)

In [ ]:
# Show my taxes!
my_taxes

So what have we learned? We learned some of the basics of numbers in Python. We also learned how to do arithmetic and use Python as a basic calculator. We then wrapped it up with learning about Variable Assignment in Python.

Up next we'll learn about Strings!

In [15]:
import this
print(this)

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
<module 'this' from 'C:\\Users\\Supriya\\anaconda3\\lib\\this.py'>
